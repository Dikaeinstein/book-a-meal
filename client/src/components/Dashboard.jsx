import React from 'react';
import OrderHistory from './OrderHistory';

const Dashboard = () => <h2>Dashboard</h2>;

export default Dashboard;
