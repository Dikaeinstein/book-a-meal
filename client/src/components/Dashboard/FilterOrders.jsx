import React from 'react';
import Button from '../util/Button';

const FilterOrders = () => (
  <div className="filter">
    From: <input type="date" />
    <span className="bar">&ThinSpace;|&ThinSpace;</span>
    To: <input type="date" />
    <Button
      className="btn btn-default font-weight-bold"
      value="Filter Orders"
    />
  </div>
);

export default FilterOrders;
