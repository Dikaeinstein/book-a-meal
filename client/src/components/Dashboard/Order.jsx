import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ConfirmationDialog from '../util/ConfirmationDialog';
import { updateOrder, deleteOrder } from '../../actions/dashboardActions';

class ConnectedOrder extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isConfirmationOpen: false,
      orderStatus: props.order.status,
    };
    this.handleDelete = this.handleDelete.bind(this);
    this.handleUpdate = this.handleUpdate.bind(this);
    this.handleCloseConfirmation = this.handleCloseConfirmation.bind(this);
  }

  handleUpdate() {
    const values = { orderStatus: this.state.orderStatus };
    this.props.modifyOrder(values);
  }

  handleDelete() {
    this.setState({ isConfirmationOpen: true });
    console.log(this.props.removeOrder);
    // this.props.removeOrder();
  }

  handleCloseConfirmation() {
    this.setState({ isConfirmationOpen: false });
  }

  render() {
    const { order } = this.props;

    const btnStyle = {
      flex: '50%',
      border: 'none',
      margin: '.5rem',
      padding: '.5rem',
      cursor: 'pointer',
      borderRadius: '3px',
      background: '#e9ebeb',
    };

    return (
      <tr key={order.id}>
        <td>{order.id}</td>
        <td>{order.meal.name}</td>
        <td>{order.user.name}</td>
        <td>{order.meal.price}</td>
        <td>{order.quantity}</td>
        <td>{order.total}</td>
        <td className="text-warning">{order.status}</td>
        <td>
          <div style={{ display: 'flex' }}>
            <button
              value="Edit"
              style={Object.assign({}, btnStyle, { color: 'green' })}
              title="Edit Order"
              onClick={this.handleUpdate}
            >
              Edit
            </button>
            <button
              value="Delete"
              style={Object.assign({}, btnStyle, { color: 'red' })}
              title="Delete Order"
              onClick={this.handleDelete}
            >
              Delete
            </button>
          </div>
        </td>
        <ConfirmationDialog
          message="Are you sure you want delete order"
          isOpen={this.state.isConfirmationOpen}
          ok={this.handleCloseConfirmation}
          cancel={this.handleCloseConfirmation}
        />
      </tr>
    );
  }
}

ConnectedOrder.propTypes = {
  /* eslint react/forbid-prop-types: 0 */
  order: PropTypes.objectOf(PropTypes.number, PropTypes.string).isRequired,
  removeOrder: PropTypes.func.isRequired,
  modifyOrder: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  removeOrder: orderId => dispatch(deleteOrder(orderId)),
  deliverOrder: (values, orderId) => dispatch(updateOrder(values, orderId)),
});

const Order = connect(null, mapDispatchToProps)(ConnectedOrder);

export default Order;
