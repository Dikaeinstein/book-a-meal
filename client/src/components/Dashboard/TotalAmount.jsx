import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getTotalAmount } from '../../actions/dashboardActions';

const ConnectedTotalAmount = ({ totalAmount, fetchTotalAmount }) => {
  fetchTotalAmount();

  return (
    <div className="dashboard-item">
      <p className="font-weight-bold">Total Amount</p>
      <h2 className="amount font-weight-bold">&#x20a6; {totalAmount}</h2>
    </div>
  );
};

ConnectedTotalAmount.propTypes = {
  totalAmount: PropTypes.number.isRequired,
  fetchTotalAmount: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  totalAmount: state.dashboard.totalAmount,
});

const mapDispatchToProps = dispatch => ({
  fetchTotalAmount: () => dispatch(getTotalAmount()),
});

const TotalAmount =
  connect(mapStateToProps, mapDispatchToProps)(ConnectedTotalAmount);

export default TotalAmount;
