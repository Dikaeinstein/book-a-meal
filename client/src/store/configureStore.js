import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';
import throttle from 'lodash.throttle';
import { loadState, saveState } from '../helpers/localStorage';
import initialState from '../reducers/initialState';
import rootReducer from '../reducers';

/**
 * Return redux middleware
 */
const getMiddleware = () => {
  const logger = createLogger();

  const middleware =
    process.env.NODE_ENV !== 'production'
    && process.env.NODE_ENV !== 'test'
      ? applyMiddleware(thunk, logger)
      : applyMiddleware(thunk);

  return middleware;
};

/**
 * Configure the redux store
 *
 * @returns {object} Redux store
 */
const configureStore = () => {
  const persistedState = loadState();

  const state = persistedState
    ? { ...initialState, ...persistedState }
    : initialState;

  const store = createStore(
    rootReducer,
    state,
    compose(
      getMiddleware(),
      window.devToolsExtension ? window.devToolsExtension() : empty => empty,
    ),
  );

  // Throttle the saveState function to be called once every second
  store.subscribe(throttle(() => {
    saveState({ orders: store.getState().orders });
  }, 1000));

  return store;
};

export default configureStore;
