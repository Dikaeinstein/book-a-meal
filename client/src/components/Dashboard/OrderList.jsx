import React from 'react';
import PropTypes from 'prop-types';
import Order from './Order';

const OrderList = ({ orders }) => {
  const renderTableBody = () => {
    const orderRows = orders.map(order => (
      <Order key={order.id} order={order} />
    ));
    return (
      <tbody>
        {orderRows}
      </tbody>
    );
  };

  return (
    <div className="orders">
      <table>
        <thead>
          <tr>
            <th>Order ID</th>
            <th>Order Name</th>
            <th>Customer</th>
            <th>Order Amount (&#x20a6;)</th>
            <th>Quantity (plate)</th>
            <th>Total (&#x20a6;)</th>
            <th>Status</th>
            <th>Order Actions</th>
          </tr>
        </thead>
        {renderTableBody()}
      </table>
    </div>
  );
};

OrderList.propTypes = {
  orders: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default OrderList;
