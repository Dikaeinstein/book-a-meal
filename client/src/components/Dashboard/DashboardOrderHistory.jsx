import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Preloader } from 'react-materialize';
import OrderList from './OrderList';
import { fetchAllOrders, fetchCatererOrders } from '../../actions/dashboardActions';
import { getOrders } from '../../reducers/dashboardReducer';
import Loading from '../util/Loading';
import Paginate from '../util/Paginate';
import {
  getNextPageUrl,
  getPreviousPageUrl, getCurrentPageUrl,
} from '../../reducers/paginationReducer';

class ConnectedDashboardOrderHistory extends Component {
  constructor(props) {
    super(props);
    this.handlePageChange = this.handlePageChange.bind(this);
    this.handleCatererPageChange = this.handlePageChange.bind(this);
  }

  componentDidMount() {
    if (this.props.role === 'superAdmin') {
      this.handlePageChange(this.props.superAdminCurrentUrl);
    } else {
      this.handleCatererPageChange(this.props.catererCurrentUrl);
    }
  }

  handleCatererPageChange(url) {
    this.props.getCatererOrders(url);
  }

  handlePageChange(url) {
    this.props.getAllOrders(url);
  }

  render() {
    const { isFetching, role } = this.props;
    if (isFetching) {
      return (
        <div className="loader-container">
          <Loading text="fetching orders...">
            <Preloader flashing size="medium" />
          </Loading>
        </div>
      );
    }

    return (
      <section>
        <h2>Order history</h2>
        <OrderList orders={this.props.allOrders} />
        <Paginate
          onPageChange={role === 'superAdmin'
            ? this.handlePageChange
            : this.handleCatererPageChange
          }
          nextUrl={role === 'superAdmin'
            ? this.props.superAdminNextUrl
            : this.props.catererNextUrl
          }
          previousUrl={role === 'superAdmin'
            ? this.props.superAdminPreviousUrl
            : this.props.catererPreviousUrl
          }
          style={{ marginTop: '0' }}
        />
      </section>
    );
  }
}

ConnectedDashboardOrderHistory.propTypes = {
  isFetching: PropTypes.bool.isRequired,
  getAllOrders: PropTypes.func.isRequired,
  getCatererOrders: PropTypes.func.isRequired,
  allOrders: PropTypes.arrayOf(PropTypes.object).isRequired,
  role: PropTypes.string.isRequired,
  superAdminCurrentUrl: PropTypes.string.isRequired,
  catererCurrentUrl: PropTypes.string.isRequired,
  superAdminNextUrl: PropTypes.string.isRequired,
  superAdminPreviousUrl: PropTypes.string.isRequired,
  catererNextUrl: PropTypes.string.isRequired,
  catererPreviousUrl: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  isFetching: state.dashboard.isFetchingAllOrders,
  allOrders: getOrders(state.dashboard),
  role: state.user.data.role,
  catererNextUrl: getNextPageUrl(state.pagination.dashboard.caterer),
  catererCurrentUrl: getCurrentPageUrl(state.pagination.dashboard.caterer),
  catererPreviousUrl: getPreviousPageUrl(state.pagination.dashboard.caterer),
  superAdminNextUrl: getNextPageUrl(state.pagination.dashboard.superAdmin),
  superAdminCurrentUrl: getCurrentPageUrl(state.pagination.dashboard.superAdmin),
  superAdminPreviousUrl: getPreviousPageUrl(state.pagination.dashboard.superAdmin),
});

const mapDispatchToProps = dispatch => ({
  getAllOrders(url) { dispatch(fetchAllOrders(url)); },
  getCatererOrders(url) { dispatch(fetchCatererOrders(url)); },
});

const DashboardOrderHistory = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ConnectedDashboardOrderHistory);

export default DashboardOrderHistory;
