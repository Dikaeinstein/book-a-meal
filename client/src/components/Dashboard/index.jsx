import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import DashboardOrderHistory from './DashboardOrderHistory';
import Footer from '../util/Footer';
import FilterOrders from './FilterOrders';
import TotalOrders from './TotalOrders';
import TotalSales from './TotalSales';
import errorHandler from '../util/errorHandler';


const ConnectedDashboard = ({ fetchOrdersError }) => {
  const DashboardOrderHistoryWithErrorHandling =
    errorHandler(DashboardOrderHistory, 'Error fetching orders');

  return (
    <div>
      <main
        className="dash-main text-center bg-dark-light"
        style={{ minHeight: 'calc(100vh - 151px)' }}
      >
        <h1>Dashboard</h1>
        <section className="card dashboard">
          <TotalOrders />
          <TotalSales />
          <FilterOrders />
          <DashboardOrderHistoryWithErrorHandling
            error={fetchOrdersError}
          />
        </section>
      </main>
      <Footer />
    </div>
  );
};

ConnectedDashboard.propTypes = {
  /* eslint react/require-default-props: 0 */
  fetchOrdersError: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.objectOf(PropTypes.string),
  ]),
};

const mapStateToProps = state => ({
  fetchOrdersError: state.orders.error,
});

const Dashboard = connect(mapStateToProps)(ConnectedDashboard);

export default Dashboard;
