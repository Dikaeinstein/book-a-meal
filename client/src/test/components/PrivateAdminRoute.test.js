import React from 'react';
import { shallow } from 'enzyme';
import { Route } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import ConnectedPrivateAdminRoute, { PrivateAdminRoute }
  from '../../components/PrivateAdminRoute';
import Dashboard from '../../components/Dashboard';

describe('<PrivateAdminRoute />', () => {
  let wrapper;
  let mockStore;
  beforeAll(() => {
    mockStore = configureStore();
  });
  it('should render with correct props', () => {
    const store = mockStore({
      user: { loggedIn: false, data: { role: 'caterer' } },
    });
    wrapper = shallow(<ConnectedPrivateAdminRoute
      path="/dashboard"
      component={Dashboard}
      store={store}
      userRole="caterer"
    />);
    expect(wrapper.props().loggedIn).toBeFalsy();
    expect(wrapper.props().path).toEqual('/dashboard');
  });
  it('should render a Component when logged in', () => {
    wrapper = shallow(<PrivateAdminRoute
      path="/dashboard"
      loggedIn
      userRole="caterer"
      component={Dashboard}
    />);
    expect(wrapper.find(Route).props().render()).toBeDefined();
  });
  it('should redirect "SignIn" when not logged in', () => {
    wrapper = shallow(<PrivateAdminRoute
      path="/dashboard"
      loggedIn={false}
      userRole="caterer"
      component={Dashboard}
    />);
    expect(wrapper.find(Route).props().render()).toBeDefined();
  });
  it('should render correctly', () => {
    wrapper = shallow(<PrivateAdminRoute
      path="/dashboard"
      loggedIn={false}
      component={Dashboard}
      userRole="caterer"
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
