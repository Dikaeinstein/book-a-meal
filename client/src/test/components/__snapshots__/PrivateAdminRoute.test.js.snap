// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrivateAdminRoute /> should render correctly 1`] = `
<Route
  path="/dashboard"
  render={[Function]}
/>
`;
